File: src/net/t3kt/tctrl/schema/SchemaNode.java
package net.t3kt.tctrl.schema;

import com.google.protobuf.Message;
import java.io.IOException;

public abstract class SchemaNode<S extends Message> {
    protected final S spec;

    protected SchemaNode(S spec) {
        this.spec = spec;
    }

    public S getSpec() {
        return spec;
    }

    public abstract String getKey();

    public abstract String getLabel();

    public abstract String getPath();

    public String toJson() {
        return Schemas.toJson(getSpec());
    }

    public void writeJson(Appendable output) throws IOException {
        Schemas.writeJson(getSpec(), output);
    }
}

File: win/TctrlCore/Schema/SchemaNode.cs
using System;
using Google.Protobuf;

namespace Tctrl.Core.Schema
{
    public interface ISchemaNode {
        string Key { get; }
        string Label { get; }
        string Path { get; }
    }
    
    public abstract class SchemaNode<S> : ISchemaNode where S : IMessage
    {
        protected SchemaNode(S spec)
        {
            Spec = spec;
        }

        public S Spec { get; }

        public abstract string Key { get; }

        public abstract string Label { get; }

        public abstract string Path { get; }
    }
}
